### PR TITLE
fix(dursto): Fix Prove and Verify Implementations of the Merkle Layer.

### PR DESCRIPTION
**Pull request by NSant215**
Closes RV-957.
Closes RV-961.

  # What

  Fix the `Prove` and `Verify` implementations of `MerkleLayer` so that proof generation
  and verification round-trip correctly across AVL-rebalancing operations
  (`set` on existing/new keys, `delete`, `write`).

  # Why

  The previous prove-mode fold walked the *working* tree, so any AVL rotation triggered
  by the step rearranged the structure relative to the proof and produced root hashes
  that no longer matched the Normal-mode hash. On the verify side, sub-proofs decoded
  into `VerifyNodeId` / `VerifyTreeId` were folded against the inherited parent proof
  rather than the sub-proof they originated from — this caused incorrect `prev_hash`
  substitutions or `PartialHash::InvalidProof` once nodes had been moved by rebalancing.

  # How

  Two structural fixes plus tests:

  1. **Verify side**: each `VerifyNodeId::Present` / `VerifyTreeId::Present` now carries
     the `MerkleProof` sub-tree it was deserialised from, captured up-front via the
     new `Deserialiser::capture_owned_proof()` hook. Their `Foldable<PartialHashFold>`
     impls override the inherited proof with this captured sub-proof before delegating,
     so each node folds against its initial-tree position even after rebalancing.
  2. **Prove side**: `start_proof()` snapshots the Normal-mode tree as an immutable
     `initial_tree`. The `MerkleProofFold` impl on `MerkleLayer<KV, Prove<'_>>` walks
     that snapshot — not the mutating working tree — using `was_node_accessed`/
     `was_tree_accessed` to decide what to blind, and pulling per-field read flags from
     the prove-mode `working_tree` / `deleted_nodes` map.

  # Per-commit breakdown

  1. Plumbing:
     - `PartialHashFold.proof` becomes `Option<Arc<MerkleProof>>` so callers can attach
       a node-local override without deep-cloning.
     - Adds `PartialHashFold::with_proof()` (replace outright, unlike
       `map_reference_proof` which requires `Some`).
     - Adds `Deserialiser::capture_owned_proof()` (default `None`; overridden by
       `ProofTree` to clone the in-memory sub-proof). Stream deserialisers can't
       reconstruct without re-parsing, so the default is correct.

  2. Use the new override:
     - `VerifyNodeId::Present` / `VerifyTreeId::Present` become struct variants
       carrying `original_proof: Option<Arc<MerkleProof>>`, captured before
       `into_node()` consumes the deserialiser.
     - Their `Foldable<PartialHashFold>` impls call `builder.with_proof(...)` before
       delegating to the inner `Node` / `Tree`, so the fold sees each node's
       own sub-proof rather than the parent's view.
     - Resolver impls and tests updated for the new struct layout.

  3. Walk the initial tree, not the mutating one:
     - Removes the broken `Foldable<MerkleProofFold>` impls on `Tree<ProveNodeId>` and
       `ProveNodeId`.
     - Adds `Foldable<MerkleProofFold>` for `MerkleLayer<KV, Prove<'_>>`, implemented
       via two wrappers (`InitialTreeFold` / `InitialNodeFold`) that recurse through the
       initial `LazyTreeId` snapshot, blinding any node/tree the resolver did not
       touch and pulling read-flag-bearing `meta` / `data` from either the working tree
       (via `Tree::get_resolved` + the new `CachedOnlyResolver`) or `deleted_nodes` for
       keys removed during the step.
     - `MerkleLayer<Normal>::start_proof()` is the new entry point: it clones the
       Normal tree, derives the prove-mode `working_tree` from it, and seeds the
       `ProveResolver` with the root hash so the root tree is always considered
       accessed (`ProveResolver::start` now takes `root_tree_hash`).
     - `ProveImpl` is now `{ initial_tree, working_tree, resolver }`. Existing
       prove-mode write-paths and tests adapted to the rename.
     - Removes `#[expect(dead_code, …)]` on `cached_node`, `inner`, `was_*_accessed`,
       `deleted_nodes`, `DeletedNodeFields` now that the fold consumes them.

  4. PBTs:
     - `assert_prove_mode_correct` asserts (1) `proof.root_hash() == initial Normal
       hash`, (2) prove-final == Normal-final after replaying the same op, (3)
       verify-mode replay over the deserialised proof reaches the same final hash.
     - Five proptest cases: random single-op (set/delete/write), writes-only,
       deletes-only, new-key insertions, and read-only.
     - Single-op cases note the current `ops_count = 1` limitation
       (multi-step proofs not yet supported).

# Manually Testing

```
make all
```

# Regressions

# Tasks for the Author

- [x] Link all Linear issues related to this MR using magic words (e.g. part of, relates to, closes).
- [x] Eliminate dead code and other spurious artefacts introduced in your changes.
- [x] Document new public functions, methods and types.
- [x] Make sure the documentation for updated functions, methods, and types is correct.
- [x] Add tests for bugs that have been fixed.
- [x] [Explain changes](#regressions) to regression test captures when applicable.
- [x] Write commit messages in agreement with our guidelines.
- [x] Self-review your changes to ensure they are high-quality.
- [x] Complete all of the above before assigning this MR to reviewers.

### CODE REVIEW

File: data/src/merkle_proof/proof_tree.rs [hunk 1]
```diff
@@ -515,6 +515,13 @@ impl<'t> Deserialiser for ProofTree<'t> {
     fn into_node(self) -> Result<Self::DeserialiserNode, Self::Error> {
         Ok(self.as_node()?.map_present(Vec::into_iter))
     }
+
+    fn capture_owned_proof(&self) -> Option<MerkleProof> {
```
Comment (emturner) [inline data/src/merkle_proof/proof_tree.rs:519]: I guess we might want to follow up on all of this, to see if we can avoid extra allocations - but works for now

File: durable-storage/src/avl/tree.rs [hunk 2]
```diff
@@ -348,6 +349,37 @@ impl<NodeId> Tree<NodeId> {
...
+impl Tree<ProveNodeId> {
+    /// Find the `ProveNodeId` and its resolved [`Node`] for a given key by traversing the tree.
+    ///
+    /// Returns `None` if the key is not present.
+    ///
+    /// Only traverses already-resolved nodes. Returns an error if an
+    /// unresolved node or tree is encountered.
+    pub(crate) fn get_resolved(
+        &self,
+        key: &Key,
```
Comment (emturner) [inline durable-storage/src/avl/tree.rs]: could this not just use `get` above? just need to pass the `CachedOnlyResolver`

File: durable-storage/src/merkle_layer.rs [hunk 3]
```diff
@@ -1975,4 +2023,264 @@ mod tests {
...
+            StepOp::Delete(key) => {
+                ml.delete(key).expect("verify delete should succeed");
+            }
+            StepOp::Write(key, data) => {
+                if ml.get(key).expect("verify get should succeed").is_some() {
+                    ml.write(key, 0, data).expect("verify write should succeed");
+                }
+            }
+        }
+    }
```
Comment (emturner) [inline durable-storage/src/merkle_layer.rs]: I'm 99% sure these three methods can be squashed into a mode-generic one?

File: durable-storage/src/merkle_layer.rs [hunk 4]
```diff
@@ -1975,4 +2023,264 @@ mod tests {
...
+            .expect("proof deserialisation should succeed")
+            .into_result();
+        let VerifyTreeId::Present {
+            tree: verify_tree, ..
+        } = verify_tree_id
+        else {
+            panic!("expected Present tree from proof");
+        };
+
+        let mut verify_ml: MerkleLayer<KV, Verify> = MerkleLayer::from_verify_tree(verify_tree);
```
Comment (emturner) [inline durable-storage/src/merkle_layer.rs]: could also assert the partial hash at this point matches the initial hash?

File: data/src/merkle_proof/proof_tree.rs (hunk 1 above)
Comment (NSant215) [reply data/src/merkle_proof/proof_tree.rs:519]: added issue [here](https://github.com/tezos/riscv-pvm/compare/60a334910fa48cb2aa29006d2cabd1e28a4af6a7..7fb176528acd2974ef66aa5fce1334988b4c77cc)

File: durable-storage/src/avl/tree.rs (hunk 2 above)
Comment (NSant215) [reply durable-storage/src/avl/tree.rs]: unfortunately not - `get` specifically only returns the `data` of the node that has the key, rather than the node id. Changed name to make that more clear [here](https://github.com/tezos/riscv-pvm/compare/7fb176528acd2974ef66aa5fce1334988b4c77cc..2c030210eafc5a3f5af94d71bc4e3fe54eb46fc8) 

I could also change the get method itself - wasn't sure what made more sense but happy either way

File: durable-storage/src/merkle_layer.rs (hunk 3 above)
Comment (NSant215) [reply durable-storage/src/merkle_layer.rs]: is this better? [diff](https://github.com/tezos/riscv-pvm/compare/2c030210eafc5a3f5af94d71bc4e3fe54eb46fc8..1f40a9351380079f1f5a114933e6b02b6c6503cd)

File: durable-storage/src/merkle_layer.rs (hunk 4 above)
Comment (NSant215) [reply durable-storage/src/merkle_layer.rs]: This made me realise that I need to implement `hash` for Verify as well, which means that the `MerkleProof` MUST be a field on Verify. Pulling back to draft whilst I work on that.

Comment (NSant215) [reply durable-storage/src/merkle_layer.rs]: done [here](https://github.com/tezos/riscv-pvm/compare/1f40a9351380079f1f5a114933e6b02b6c6503cd..9278f76f74e92f46f04572d24057290379644e17)

File: durable-storage/src/merkle_layer.rs (hunk 3 above)
Comment (emturner) [reply durable-storage/src/merkle_layer.rs]: I think just something like this is okay? Is there any particular reason we have to do the `get` first?
```
    fn apply_step_op<KV: KeyValueStore, M: MerkleLayerMode>(ml: &mut MerkleLayer<KV, M>, op: &StepOp) {
        match op {
            StepOp::Set(key, data) => {
                ml.set(key, data).expect("set should succeed");
            }
            StepOp::Delete(key) => {
                ml.delete(key).expect("delete should succeed");
            }
            StepOp::Write(key, data) => {
                ml.write(key, 0, data).expect("write should succeed");
            }
            // TODO: add `get` to step op
        }
    }
```

(which, really becomes 'two steps' for the write at that point)

Comment (NSant215) [reply durable-storage/src/merkle_layer.rs]: lost diff but done now - this change highlighted a bug where writing a new Node would fail the test, as the tree itself wouldn't be accessed, and therefore would be blinded in the proof, and wouldn't be able to later be made present.

File: durable-storage/src/avl/tree.rs [hunk 5]
```diff
@@ -47,6 +48,16 @@ use crate::storage::Loadable;
 use crate::storage::Storable;
 use crate::storage::StoreOptions;
 
+/// Hash of the empty AVL tree (`Tree::<NodeId>(None)`).
+/// The empty-tree hash is independent of the `NodeId` type parameter.
+static EMPTY_TREE_HASH: LazyLock<Hash> = LazyLock::new(|| {
+    Hash::from([
```
Comment (emturner) [inline durable-storage/src/avl/tree.rs]: this shouldn't be hardcoded - this can be calculated directly

File: durable-storage/src/merkle_layer.rs (hunk 3 above)
Comment (emturner) [reply durable-storage/src/merkle_layer.rs]: fixed offline

File: durable-storage/src/avl/tree.rs [hunk 6]
```diff
@@ -888,6 +896,16 @@ mod tests {
...
+    fn empty_tree_hash_constant_matches_default() {
+        assert_eq!(
+            *EMPTY_TREE_HASH,
+            Hash::from_foldable(&Tree::<LazyNodeId>(None)),
+            "EMPTY_TREE_HASH literal is stale — recompute it from \
+             `Hash::from_foldable(&Tree::<LazyNodeId>(None))` and update the byte array.",
+        );
+    }
+
     #[test]
```
Comment (kurtisc) [inline durable-storage/src/avl/tree.rs]: Not needed now

File: durable-storage/src/merkle_layer.rs [hunk 7]
```diff
@@ -83,6 +89,37 @@ impl<KV> MerkleLayer<KV, Normal> {
...
+    ///
+    /// The proof is retained so [`MerkleLayer::hash`] can resolve `prev_hash` substitutions for
+    /// blinded subtrees that fold to [`PartialHash::Previous`].
+    #[cfg_attr(
+        not(test),
+        expect(
+            dead_code,
+            reason = "non-test callers wired in follow-up PVM integration (RV-957)"
+        )
+    )]
```
Comment (kurtisc) [inline durable-storage/src/merkle_layer.rs]: There is already a test module in this file, you may find it cleaner to simply move this function into it, then move it out after RV-957

File: durable-storage/src/merkle_layer.rs [hunk 8]
```diff
@@ -250,12 +287,22 @@ impl MerkleLayerMode for Verify {
...
                 resolver: VerifyResolver,
+                proof: this.inner.proof.clone(),
             },
         }
     }
 
-    fn hash<KV>(_this: &MerkleLayer<KV, Self>) -> Hash {
-        unimplemented!("Blocked by RV-961")
+    fn hash<KV>(this: &MerkleLayer<KV, Self>) -> Hash {
+        // The wrapping `original_proof: None` is intentional: it lets each per-node sub-proof
```
Comment (kurtisc) [inline durable-storage/src/merkle_layer.rs]: ~~So make it explicit?~~

Should be removed now

File: durable-storage/src/merkle_layer.rs [hunk 9]
```diff
@@ -400,15 +450,20 @@ impl<KV> MerkleLayer<KV, Verify> {
...
 }
 
+/// Construct a verify-mode [`MerkleLayer`] for an empty initial state. Used in tests that exercise
+/// verify-mode mutation primitives in isolation.
+#[cfg(test)]
+pub(crate) fn empty_verify_layer<KV>() -> MerkleLayer<KV, Verify> {
+    let proof = MerkleProof::from_foldable(&Tree::<ProveNodeId>::default());
+    MerkleLayer::from_proof(proof).expect("The proof should be deserialisable")
+}
+
```
Comment (kurtisc) [inline durable-storage/src/merkle_layer.rs:608]: Move to module below

File: durable-storage/src/merkle_layer.rs [hunk 10]
```diff
@@ -400,15 +450,20 @@ impl<KV> MerkleLayer<KV, Verify> {
     }
 }
 
+/// Construct a verify-mode [`MerkleLayer`] for an empty initial state. Used in tests that exercise
+/// verify-mode mutation primitives in isolation.
+#[cfg(test)]
+pub(crate) fn empty_verify_layer<KV>() -> MerkleLayer<KV, Verify> {
```
Comment (kurtisc) [inline durable-storage/src/merkle_layer.rs]: `new_verify_layer` for consistency with the other helpers?

File: durable-storage/src/avl/tree.rs (hunk 2 above)
Comment (kurtisc) [reply durable-storage/src/avl/tree.rs]: Would it be possible to swap `get` to something that returns the ID, then have another method on the ID to get the data?

File: durable-storage/src/merkle_layer.rs [hunk 11]
```diff
@@ -1798,4 +2052,216 @@ mod tests {
...
+    /// 2. The prove-mode final hash matches Normal-mode after an identical operation (prove-mode
+    ///    mutations are faithful).
+    /// 3. Replaying the same operation against the Verify-mode tree decoded from the proof
+    ///    produces the same final hash as prove mode (the proof carries enough information for
+    ///    full verification).
+    fn assert_prove_mode_correct<KV: KeyValueStore + TestKeyValueStoreSetup>(
+        setup_keys: &[[u8; 2]],
+        step_ops: &[StepOp],
+    ) {
+        // ---- Normal: build initial tree ----
```
Comment (kurtisc) [inline durable-storage/src/merkle_layer.rs]: Humanise pls

File: durable-storage/src/merkle_layer.rs [hunk 12]
```diff
@@ -2003,4 +2052,216 @@ mod tests {
...
+            }
+            StepOp::Delete(key) => {
+                ml.delete(key).expect("delete should succeed");
+            }
+            StepOp::Write(key, data) => {
+                ml.write(key, 0, data).expect("write should succeed");
+            } // TODO RV-895: add `get` to step op
+        }
+    }
+    /// Core assertion: prove-mode proof and hash are consistent with Normal mode and the proof
```
Comment (kurtisc) [inline durable-storage/src/merkle_layer.rs]: Formatting error. Also, it is preferable (to me) for the helper functions to be located at the top of the tests module with the others, not interspersed with tests

File: durable-storage/src/merkle_layer.rs [hunk 13]
```diff
@@ -2003,4 +2052,216 @@ mod tests {
...
+                    0 => step_ops.push(StepOp::Set(key, data)),
+                    1 => step_ops.push(StepOp::Delete(key)),
+                    _ => step_ops.push(StepOp::Write(key, data)),
+                }
+            }
+
+            assert_prove_mode_correct::<KV>(&setup_keys, &step_ops);
+        });
+
+        // prove_verify_writes_only (no structural change)
```
Comment (kurtisc) [inline durable-storage/src/merkle_layer.rs]: Explain why these don't cause structural changes.

The underscore item-like prefixes don't seem to add much to these comments

File: durable-storage/src/merkle_layer.rs [hunk 14]
```diff
@@ -2003,4 +2052,216 @@ mod tests {
...
+        proptest!(|(setup_keys in prop::collection::vec(any::<[u8; 2]>(), 1..30))| {
+            let keys: Vec<Key> = setup_keys
+                .iter()
+                .map(|b| Key::new(b).expect("key should be valid"))
+                .collect();
+
+            let step_ops: Vec<StepOp> = keys
+                .iter()
+                .enumerate()
+                .map(|(i, key)| StepOp::Write(key.clone(), vec![0xAA; 1 + (i % 8)]))
```
Comment (kurtisc) [inline durable-storage/src/merkle_layer.rs]: Is there any significance to this pattern? Resolve if not

**Review by kurtisc — CHANGES_REQUESTED** [review]

File: durable-storage/src/avl/tree.rs (hunk 2 above)
Comment (kurtisc) [reply durable-storage/src/avl/tree.rs]: This was resolved by Emma, re-resolve if the suggestion doesn't work

File: durable-storage/src/merkle_layer.rs [hunk 15]
```diff
@@ -391,6 +440,9 @@ struct ProveImpl<KV> {
 struct VerifyImpl {
     tree: Tree<VerifyNodeId>,
     resolver: VerifyResolver,
+    /// Original proof the verify-mode tree was deserialised from. Required by [`MerkleLayer::hash`]
+    /// to resolve `prev_hash` substitutions in [`PartialHashFold::previous`] for blinded subtrees.
+    proof: Arc<MerkleProof>,
```
Comment (emturner) [inline durable-storage/src/merkle_layer.rs]:
```suggestion
    original_proof: Arc<MerkleProof>,
```

File: durable-storage/src/merkle_layer.rs (hunk 9 above)
Comment (kurtisc) [reply durable-storage/src/merkle_layer.rs:608]: Used elsewhere

File: durable-storage/src/merkle_layer.rs [hunk 16]
```diff
@@ -2005,4 +2054,216 @@ mod tests {
...
+        op: &StepOp,
+    ) {
+        match op {
+            StepOp::Set(key, data) => {
+                ml.set(key, data).expect("set should succeed");
+            }
+            StepOp::Delete(key) => {
+                ml.delete(key).expect("delete should succeed");
+            }
+            StepOp::Write(key, data) => {
```
Comment (NSant215) [inline durable-storage/src/merkle_layer.rs]: add test for non-zero offset writes

File: durable-storage/src/avl/tree.rs (hunk 2 above)
Comment (NSant215) [reply durable-storage/src/avl/tree.rs]: done [here](https://github.com/tezos/riscv-pvm/compare/bb108e9..ca8368c62fcc638f4f917a843499d0bba809fed1)

File: durable-storage/src/merkle_layer.rs (hunk 7 above)
Comment (NSant215) [reply durable-storage/src/merkle_layer.rs]: done [here](https://github.com/tezos/riscv-pvm/compare/ca8368c62fcc638f4f917a843499d0bba809fed1..c4b802a893196d56fdec8b55312023fdbe4cf2a4)

File: durable-storage/src/avl/tree.rs (hunk 6 above)
Comment (NSant215) [reply durable-storage/src/avl/tree.rs]: lost diff but gone I think

File: durable-storage/src/merkle_layer.rs (hunk 8 above)
Comment (NSant215) [reply durable-storage/src/merkle_layer.rs]: lost diff but think I have done now

File: durable-storage/src/merkle_layer.rs (hunk 10 above)
Comment (NSant215) [reply durable-storage/src/merkle_layer.rs]: done [here](https://github.com/tezos/riscv-pvm/compare/aa3f044efd392e4870cad164da863fda899c5e71..f59ac40769316a5f244a00783160264c3792f099)

File: durable-storage/src/merkle_layer.rs (hunk 15 above)
Comment (NSant215) [reply durable-storage/src/merkle_layer.rs]: done [here](https://github.com/tezos/riscv-pvm/compare/859ac24edb6fdedec014332d4bfe5722cc3f20e5..21a6a1827a2f76c693423552fd37d37bf5025acd)

File: durable-storage/src/merkle_layer.rs (hunk 11 above)
Comment (NSant215) [reply durable-storage/src/merkle_layer.rs]: I find the separation useful? Can you elaborate XD

File: durable-storage/src/merkle_layer.rs (hunk 13 above)
Comment (NSant215) [reply durable-storage/src/merkle_layer.rs]: fixed [here](https://github.com/tezos/riscv-pvm/compare/27f7ac2d2a061d877050de1010103705389e90a6..03da16a609d3abf9046755f7e11749dcb2981937)

File: durable-storage/src/merkle_layer.rs (hunk 12 above)
Comment (NSant215) [reply durable-storage/src/merkle_layer.rs]: moved (lost diff)

File: durable-storage/src/merkle_layer.rs (hunk 16 above)
Comment (NSant215) [reply durable-storage/src/merkle_layer.rs]: done [here](https://github.com/tezos/riscv-pvm/compare/03da16a609d3abf9046755f7e11749dcb2981937..5f4f0bb1b7cb4289ddcd8c892ed138267bddfb81)

**Review by emturner — APPROVED** [review]

File: durable-storage/src/avl/node.rs [hunk 17]
```diff
@@ -134,20 +143,21 @@ impl<TreeId, M: Mode> Node<TreeId, M> {
...
     where
         NodeId: 'a,
         TreeId: 'a,
+        M: 'a,
     {
         loop {
             let resolved_node = resolver.resolve(node)?;
             match resolved_node.key().cmp(key) {
-                Ordering::Equal => return Ok(Some(&resolved_node.data)),
+                Ordering::Equal => return Ok(Some(node)),
```
Comment (kurtisc) [inline durable-storage/src/avl/node.rs:160]: One more thing, please break out these changes (and anything else that cascades) into their own commit

**Review by kurtisc — APPROVED** [review]

File: durable-storage/src/avl/node.rs (hunk 17 above)
Comment (NSant215) [reply durable-storage/src/avl/node.rs:160]: done [here](https://github.com/tezos/riscv-pvm/compare/25ef8e42a4fc7cc07fa6832017b5a92651fb1298..d3e6305e269eb4cb40a3071ae2508f0bad22b9ed)
